pkg/traces/promsdprocessor/consumer/consumer.go

```diff
@@ -0,0 +1,254 @@
+	OperationTypeInsert = "insert"
+	// OperationTypeUpdate only modifies an existing k/v
+	OperationTypeUpdate = "update"
+	// OperationTypeUpsert does both of above
+	OperationTypeUpsert = "upsert"
+
+	//TODO: It'd be cleaner to get these from the otel semver package?
+	//      Not all are in semver though. E.g. "k8s.pod.ip" is internal inside the k8sattributesprocessor.
+	PodAssociationIPLabel       = "ip"
+	PodAssociationOTelIPLabel   = "net.host.ip"
+	PodAssociationk8sIPLabel    = "k8s.pod.ip"
+	PodAssociationHostnameLabel = "hostname"
+	PodAssociationConnectionIP  = "connection"
+)
+
+func ValidateOperationType(operationType string) error {
+	switch operationType {
+	case
+		OperationTypeUpsert,
+		OperationTypeInsert,
+		OperationTypeUpdate:
+		// Valid configuration, do nothing.
+	default:
+		return fmt.Errorf("unknown operation type %s", operationType)
+	}
+	return nil
+}
+
+func ValidatePodAssociations(podAssociations []string) error {
+	for _, podAssociation := range podAssociations {
+		switch podAssociation {
+		case
+			PodAssociationIPLabel,
+			PodAssociationOTelIPLabel,
+			PodAssociationk8sIPLabel,
+			PodAssociationHostnameLabel,
+			PodAssociationConnectionIP:
+			// Valid configuration, do nothing.
+		default:
+			return fmt.Errorf("unknown pod association %s", podAssociation)
+		}
+	}
+	return nil
+}
+
+// TODO: Put a private member so that this can't be created without calling NewConsumer?
+type Consumer struct {
+	optsMut sync.RWMutex
+	opts    Options
+
+	logger log.Logger
+}
+
+// Options configure a Consumer
+type Options struct {
+	HostLabels      map[string]discovery.Target
+	OperationType   string
+	PodAssociations []string
+	NextConsumer    otelconsumer.Traces
+}
+
+var _ otelconsumer.Traces = (*Consumer)(nil)
+
+func NewConsumer(nextConsumer otelconsumer.Traces, operationType string, podAssociations []string, logger log.Logger) (*Consumer, error) {
+	if nextConsumer == nil {
+		return nil, otelcomponent.ErrNilNextConsumer
+	}
+
+	err := ValidateOperationType(operationType)
+	if err != nil {
+		return nil, err
+	}
+
+	err = ValidatePodAssociations(podAssociations)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Consumer{
+		opts: Options{
+			HostLabels:      make(map[string]discovery.Target),
+			OperationType:   operationType,
+			PodAssociations: podAssociations,
+			NextConsumer:    nextConsumer,
+		},
+		logger: logger,
+	}, nil
+}
+
+// UpdateOptions is used in flow mode, where all options need to be updated.
+func (c *Consumer) UpdateOptions(opts Options) {
+	c.optsMut.Lock()
+	defer c.optsMut.Unlock()
+
+	c.opts = opts
+}
+
+// UpdateOptionsHostLabels is used in static mode, where only the host labels need to be updated.
+func (c *Consumer) UpdateOptionsHostLabels(hostLabels map[string]discovery.Target) {
+	c.optsMut.Lock()
+	defer c.optsMut.Unlock()
+
+	c.opts.HostLabels = hostLabels
+}
+
+func (c *Consumer) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	c.optsMut.RLock()
```

Comment:
@mattdurham  I also moved this mutex here instead of in `processAttributes`. This is so that we use a consistent set of "next consumer" and data from the discovery components. Unfortunately this means the mutex is held unnecessarily long for static mode, because out of the four "options" static mode only updates HostLabels:

* HostLabels
* OperationType 
* PodAssociations
* NextConsumer

I think this should be fine because I expect this traces processing to complete in less time than Prometheus' service discovery refresh. But want to give you a heads up.